diff --git a/drivers/block/blktap2/blktap.h b/drivers/block/blktap2/blktap.h
index 591bc88..49d4bdf 100644
--- a/drivers/block/blktap2/blktap.h
+++ b/drivers/block/blktap2/blktap.h
@@ -24,7 +24,7 @@ extern int blktap_device_major;
 #define BTWARN(_f, _a...)            BTPRINTK(0, KERN_WARNING, 0, _f, ##_a)
 #define BTERR(_f, _a...)             BTPRINTK(0, KERN_ERR, 0, _f, ##_a)
 
-#define MAX_BLKTAP_DEVICE            2048
+#define MAX_BLKTAP_DEVICE            4096
 #define BLKTAP_DEVICE                4
 #define BLKTAP_DEVICE_CLOSED         5
 #define BLKTAP_SHUTDOWN_REQUESTED    8
diff --git a/drivers/block/blktap2/control.c b/drivers/block/blktap2/control.c
index b35e1a7..e0cc82c 100644
--- a/drivers/block/blktap2/control.c
+++ b/drivers/block/blktap2/control.c
@@ -35,22 +35,6 @@ blktap_control_get_minor(void)
 	if (minor == MAX_BLKTAP_DEVICE)
 		goto fail;
 
-	if (minor == blktap_max_minor) {
-		void *p;
-		int n;
-
-		n = min(2 * blktap_max_minor, MAX_BLKTAP_DEVICE);
-		p = krealloc(blktaps, n * sizeof(blktaps[0]), GFP_KERNEL);
-		if (!p)
-			goto fail;
-
-		blktaps          = p;
-		minor            = blktap_max_minor;
-		blktap_max_minor = n;
-
-		memset(&blktaps[minor], 0, (n - minor) * sizeof(blktaps[0]));
-	}
-
 	tap->minor = minor;
 	blktaps[minor] = tap;
 
@@ -249,7 +233,7 @@ blktap_control_init(void)
 
 	control_device = blktap_control.this_device;
 
-	blktap_max_minor = min(64, MAX_BLKTAP_DEVICE);
+	blktap_max_minor = MAX_BLKTAP_DEVICE;
 	blktaps = kzalloc(blktap_max_minor * sizeof(blktaps[0]), GFP_KERNEL);
 	if (!blktaps) {
 		BTERR("failed to allocate blktap minor map");
