diff -r 0112273cb9db drivers/block/blktap/device.c
--- a/drivers/block/blktap/device.c	Tue May 03 12:34:29 2011 -0700
+++ b/drivers/block/blktap/device.c	Wed Nov 30 14:15:17 2011 +0000
@@ -1,3 +1,4 @@
+#include <linux/export.h>
 #include <linux/fs.h>
 #include <linux/blkdev.h>
 #include <linux/cdrom.h>
diff -r 0112273cb9db drivers/block/blktap/ring.c
--- a/drivers/block/blktap/ring.c	Tue May 03 12:34:29 2011 -0700
+++ b/drivers/block/blktap/ring.c	Wed Nov 30 14:15:17 2011 +0000
@@ -1,5 +1,6 @@
 
 #include <linux/device.h>
+#include <linux/export.h>
 #include <linux/signal.h>
 #include <linux/sched.h>
 #include <linux/poll.h>
