diff --git a/mm/page_alloc.c b/mm/page_alloc.c
index 494a081..55f7b11 100644
--- a/mm/page_alloc.c
+++ b/mm/page_alloc.c
@@ -952,6 +952,13 @@ int move_freepages(struct zone *zone,
 	 * Remove at a later date when no bug reports exist related to
 	 * grouping pages by mobility
 	 */
+	if(page_zone(start_page) != page_zone(end_page)) {
+		printk(KERN_CRIT
+		       "%s(): zone %s for %lu != zone %s for %lu\n",
+		       __func__,
+		       page_zone(start_page)->name, page_to_pfn(start_page),
+		       page_zone(end_page)->name, page_to_pfn(end_page));
+	}
 	BUG_ON(page_zone(start_page) != page_zone(end_page));
 #endif
 
