diff --git a/drivers/xen/events/events_base.c b/drivers/xen/events/events_base.c
index d91551d..8e2ccf6 100644
--- a/drivers/xen/events/events_base.c
+++ b/drivers/xen/events/events_base.c
@@ -1112,6 +1112,15 @@ void unbind_from_irqhandler(unsigned int irq, void *dev_id)
 
 	if (WARN_ON(!info))
 		return;
+
+	{
+		struct irq_desc *desc;
+
+		desc = irq_to_desc(irq);
+		pr_info("%s(%d): %s %s\n", __func__, irq, desc->name,
+			desc->action ? desc->action->name : "(none)");
+	}
+
 	free_irq(irq, dev_id);
 	unbind_from_irq(irq);
 }
diff --git a/kernel/irq/handle.c b/kernel/irq/handle.c
index 131ca17..fff5361 100644
--- a/kernel/irq/handle.c
+++ b/kernel/irq/handle.c
@@ -135,6 +135,11 @@ handle_irq_event_percpu(struct irq_desc *desc, struct irqaction *action)
 	irqreturn_t retval = IRQ_NONE;
 	unsigned int flags = 0, irq = desc->irq_data.irq;
 
+	if (!action) {
+		pr_info("irq %d NULL action\n", irq);
+		/* Let it crash so XenRT will spot it. */
+	}
+
 	do {
 		irqreturn_t res;
 
