diff --git a/arch/x86/include/asm/xen/hypercall.h b/arch/x86/include/asm/xen/hypercall.h
index be81673..a0b1799 100644
--- a/arch/x86/include/asm/xen/hypercall.h
+++ b/arch/x86/include/asm/xen/hypercall.h
@@ -458,6 +458,13 @@ HYPERVISOR_hvm_op(int op, void *arg)
        return _hypercall2(unsigned long, hvm_op, op, arg);
 }
 
+static inline int __must_check
+HYPERVISOR_kexec_op(
+        unsigned long op, void *args)
+{
+	return _hypercall2(int, kexec_op, op, args);
+}
+
 static inline int
 HYPERVISOR_tmem_op(
 	struct tmem_op *op)
diff --git a/include/xen/interface/kexec.h b/include/xen/interface/kexec.h
new file mode 100644
index 0000000..cba9b77
--- /dev/null
+++ b/include/xen/interface/kexec.h
@@ -0,0 +1,9 @@
+#ifndef _XEN_KEXEC_H
+#define _XEN_KEXEC_H
+
+#define KEXEC_CMD_kexec                    0
+typedef struct xen_kexec_exec {
+	int type;
+} xen_kexec_exec_t;
+
+#endif
diff --git a/include/xen/interface/xen.h b/include/xen/interface/xen.h
index 34adebe..04e5ce8 100644
--- a/include/xen/interface/xen.h
+++ b/include/xen/interface/xen.h
@@ -57,6 +57,7 @@
 #define __HYPERVISOR_event_channel_op     32
 #define __HYPERVISOR_physdev_op           33
 #define __HYPERVISOR_hvm_op               34
+#define __HYPERVISOR_kexec_op             37
 #define __HYPERVISOR_tmem_op              38
 
 /* Architecture-specific hypercall definitions. */
diff --git a/kernel/panic.c b/kernel/panic.c
index 167ec09..c9ed5e8 100644
--- a/kernel/panic.c
+++ b/kernel/panic.c
@@ -23,6 +23,12 @@
 #include <linux/init.h>
 #include <linux/nmi.h>
 
+#ifdef CONFIG_XEN
+#include <xen/xen.h>
+#include <asm/xen/hypercall.h>
+#include <xen/interface/kexec.h>
+#endif
+
 #define PANIC_TIMER_STEP 100
 #define PANIC_BLINK_SPD 18
 
@@ -108,6 +114,12 @@ void panic(const char *fmt, ...)
 		dump_stack();
 #endif
 
+#ifdef CONFIG_XEN
+	if (xen_initial_domain()) {
+		xen_kexec_exec_t xke = { KEXEC_TYPE_CRASH };
+		HYPERVISOR_kexec_op(KEXEC_CMD_kexec, &xke);
+	}
+#endif
 	/*
 	 * If we have crashed and we have a crash kernel loaded let it handle
 	 * everything else.
