diff --git a/drivers/xen/events/events_base.c b/drivers/xen/events/events_base.c
index d91551d..000cd3b 100644
--- a/drivers/xen/events/events_base.c
+++ b/drivers/xen/events/events_base.c
@@ -367,7 +367,7 @@ void notify_remote_via_irq(int irq)
 	if (VALID_EVTCHN(evtchn))
 		notify_remote_via_evtchn(evtchn);
 }
-EXPORT_SYMBOL_GPL(notify_remote_via_irq);
+EXPORT_SYMBOL(notify_remote_via_irq);
 
 static void xen_irq_init(unsigned irq)
 {
@@ -1062,7 +1062,7 @@ int bind_interdomain_evtchn_to_irqhandler(unsigned int remote_domain,
 
 	return irq;
 }
-EXPORT_SYMBOL_GPL(bind_interdomain_evtchn_to_irqhandler);
+EXPORT_SYMBOL(bind_interdomain_evtchn_to_irqhandler);
 
 int bind_virq_to_irqhandler(unsigned int virq, unsigned int cpu,
 			    irq_handler_t handler,
@@ -1115,7 +1115,7 @@ void unbind_from_irqhandler(unsigned int irq, void *dev_id)
 	free_irq(irq, dev_id);
 	unbind_from_irq(irq);
 }
-EXPORT_SYMBOL_GPL(unbind_from_irqhandler);
+EXPORT_SYMBOL(unbind_from_irqhandler);
 
 /**
  * xen_set_irq_priority() - set an event channel priority.
