diff --git a/drivers/net/xen-netback/netback.c b/drivers/net/xen-netback/netback.c
index da85ffb..a55cb52 100644
--- a/drivers/net/xen-netback/netback.c
+++ b/drivers/net/xen-netback/netback.c
@@ -596,8 +596,8 @@ static void xenvif_rx_action(struct xenvif *vif)
 			 * number of times xenvif_gop_frag_copy() will
 			 * call start_new_rx_buffer().
 			 */
-			max_slots_needed += DIV_ROUND_UP(offset + size,
-							 PAGE_SIZE);
+			max_slots_needed += DIV_ROUND_UP(offset_in_page(offset)
+							+ size, PAGE_SIZE);
 		}
 
 		/* To avoid the estimate becoming too pessimal for some
