diff --git a/drivers/net/xen-netback/netback.c b/drivers/net/xen-netback/netback.c
index 678bc55..4aee727 100644
--- a/drivers/net/xen-netback/netback.c
+++ b/drivers/net/xen-netback/netback.c
@@ -1761,11 +1761,11 @@ int xenvif_kthread(void *data)
 		if (kthread_should_stop())
 			break;
 
+		vif->rx_event = false;
+
 		if (!skb_queue_empty(&vif->rx_queue))
 			xenvif_rx_action(vif);
 
-		vif->rx_event = false;
-
 		if (skb_queue_empty(&vif->rx_queue) &&
 		    netif_queue_stopped(vif->dev))
 			xenvif_start_queue(vif);
