# HG changeset patch
# Parent eca3a2e7ccdf203340c1f480e0086517448886be

diff -r eca3a2e7ccdf arch/x86/pci/pcbios.c
--- a/arch/x86/pci/pcbios.c	Tue Apr 03 14:47:15 2012 +0100
+++ b/arch/x86/pci/pcbios.c	Thu Apr 05 13:07:12 2012 +0100
@@ -447,7 +447,7 @@ EXPORT_SYMBOL(pcibios_set_irq_routing);
 
 void __init pci_pcbios_init(void)
 {
-	if ((pci_probe & PCI_PROBE_BIOS) 
+	if ((pci_probe & PCI_PROBE_BIOS) && !xen_domain()
 		&& ((raw_pci_ops = pci_find_bios()))) {
 		pci_bios_present = 1;
 	}
